### DISCUSSION TITLE
How to enforce configuration of two alternative second factors (TOTP and Recovery Codes)

### DISCUSSION THREAD

**Discussion by jbman** [keycloak/keycloak#19548] [Q&A]
We want to have step-up two-factor authentication, so that
- the user is required to setup a TOTP authenticator application AND recovery codes. 
- the user can select between both factors on authentication: TOTP is used if available and recovery is only used when device isn't available or lost
 
The following authentication flow works fine, as long as the user has pre-setup the factors via Account UI. 

```
Cookie (Alternative)
Identity Provider Redirector (Alternative)
Flow "Step Up and Recovery Forms" (Alternative)
- Username Password Form (Required)
- Conditional Flow "Level 1" (Conditional)
   - Conditional - Level of Authentication (Required), , Config: loa-condition-level: 1
   - Allow Access (Required)
- Conditional Flow "Level 2" (Conditional)
   - Conditional - Level of Authentication (Required), Config: loa-condition-level: 2
   - Flow "OTP or Recovery Codes" (Required)
     - OTP Form (Alternative)
     - Recovery Authentication Form (Alternative)
```

With this flow and URL param **acr_values=2** the following happens, depending on already existing configuration of TOTP and Recovery Codes at the user:
| TOTP Configured  | Recovery Codes configured | Result |
|----------------------------|------------------------------------|----------|
| no                             | no                                   | "Error: Invalid Username or Password" is displayed |
| yes                           | no                                    | User can select between TOTP and Recovery codes. When recovery code is selected: "Error: Cannot login, credential setup required." is displayed |
| no                            | yes                                    | A recovery code is requested |

What is needed: Setup of TOTP and Recovery codes should be completed in any case.

Is there some other flow setup which enforces initial configuration of both factors on first use AND allows the user to select the factor on login?

If not, would it be an option to make the "OTP Form" executor and "Recovery Authentication Form" executor configurable? (So that configuration can be set to "required" / "not required" and isn't purely depending on Alternative/Required setting at authentication time?)

Related question: 
- https://github.com/keycloak/keycloak/discussions/17180
- https://github.com/keycloak/keycloak/discussions/14988 - This solution only enforces setup of one of the second factors

**Comment by stianst** [comment]
@mposolda could you take a look at this one?

**Comment by jbman** [comment]
When an Authenticator is set to "Alternative" the code at [DefaultAuthenticationFlow.java#L425](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java#L425) doesn't call `Authenticator.setRequiredActions(session)`.
We need a way to set the `CONFIGURE_...` action even if the authenticator is "Alternative" for the authentication flow.

**Comment by eggyknap** [comment]
I'd love to see how to do this, as well. Our users hope to log in with either OTP or WebAuthn, with recovery codes as a fallback. When I logged into github just now to post this comment, my much-abused USB security key didn't work, and conveniently, github had a set of links ready for me to choose SMS or OTP as alternative second factors. I'd like to do something like that, and don't see how.

**Reply by eggyknap** [reply]
My plan as it stands now, with inspiration from [this post](https://github.com/keycloak/keycloak/discussions/14988), is to have "recovery authentication codes" as a default required action for the realm, and to modify the default browser authentication flow to disable the "Condition - user configured" preceding "OTP Form", to change "OTP Form" to "alternative", and to add "WebAuthn authenticator" and "Recovery Authentication Code Form" as additional alternatives in that subflow. This results in my 21.0.1 installation in the user being prompted to use OTP, WebAuthn, or Recovery codes. The system selects the first one of those factors that's configured for the user, and presents them with a link to "Try another way", which lets them choose a different factor instead.

I don't have a good way yet for users to choose whether they prefer OTP or WebAuthn, but I expect in our case that administrators will be able to know when they create a new user which of the two is most appropriate, and to add to the user the proper required action to configure the user's choice of second factor.
